fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Quantifier.java

```diff
@@ -0,0 +1,224 @@
+ * consumption of that data by another {@link RelationalExpression} S. S is said to own the quantifier, while the
+ * quantifier is said to range over R. Quantifiers come in very few but very distinct flavors. All flavors are
+ * implemented by static inner final classes as to emulate a sealed trait.
+ *
+ * Quantifiers separate what it means to be producing versus consuming records. The expression a quantifier ranges over
+ * produces records, the quantifier flows information (according to the flavor) which is then consumed by the expression
+ * containing or owning the quantifier. That expression can consume the data in a way independent of how the data was
+ * produced in the first place.
+ *
+ * A quantifier works closely with the expression that owns it. Depending on the semantics of the owning expression
+ * it becomes possible to model correlations,  e.g. for a logical join expression the quantifier can provide a binding
+ * of the record being currently consumed by the join's outer to other (inner) parts of the data flow that are also rooted
+ * at the owning (join) expression.
+ */
+public abstract class Quantifier implements Bindable {
+    /**
+     * For Each quantifier. Conceptually flows one record at a time from the expression it ranges over to
+     * the owning expression.
+     */
+    public static final class ForEach extends Quantifier {
+        private final ExpressionRef<? extends RelationalExpression> rangesOver;
+
+        private ForEach(ExpressionRef<? extends RelationalExpression> rangesOver) {
+            this.rangesOver = rangesOver;
+        }
+
+        @Override
+        @Nonnull
+        public ExpressionRef<? extends RelationalExpression> getRangesOver() {
+            return rangesOver;
+        }
+
+        @Override
+        @Nonnull
+        public String getShorthand() {
+            return "ƒ";
+        }
+    }
+
+    /**
+     * Factory method to create a for-each quantifier over a given expression reference containing relational
+     * expressions.
+     * @param rangesOver expression reference to {@link RelationalExpression}s
+     * @return a for-each quantifier ranging over the given expression reference
+     */
+    @Nonnull
+    public static ForEach forEach(final ExpressionRef<? extends RelationalExpression> rangesOver) {
+        return new ForEach(rangesOver);
+    }
+
+    /**
+     * Existential quantifier. Conceptually flows exactly one record containing a boolean to the owning
+     * expression indicating whether the sub graph the quantifier ranges over produced a non-empty or an empty
+     * result. When the semantics of this quantifiers are realized in an execution strategy that strategy should
+     * facilitate an early out mechanism as the result will be {@code record(true)} as soon as the sub graph produces
+     * the first record.
+     */
+    @SuppressWarnings("squid:S2160") // sonarqube thinks .equals() and heshCode() should be overwritten which is not necessary
```

Comment:
Any idea why `ForEach` is not an issue?
```suggestion
    @SuppressWarnings("squid:S2160") // sonarqube thinks .equals() and hashCode() should be overwritten which is not necessary
```